9945 java17 nashorn

mposolda opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mposolda committed:
Nashorn working on KeycloakServer app

mposolda committed:
Simple script authenticator working on Wildfly and Quarkus

DGuhr commented on common/src/main/java/org/keycloak/common/util/Environment.java:
```diff
@@ -26,6 +26,21 @@ public class Environment {
+    /*
+     * @return major version of java runtime like 8, 11 or 17
+     */
+    public static int getJavaVersion() {
```
when this is targeted to the time we won't have support for wildfly/java8 anymore, I suggest to use `Runtime.version().major()` instead. way less verbose and gives what you want in one line of code. Available since java9.

mposolda replied:
I've changed the quarkus part per the suggestion of @pedroigor so it doesn't use this anymore. Hence I removed the stuff from the `Environment` class of keycloak-common module and moved it to the Wildfly module.

However Wildfly module still needs to use that instead of `Runtime.version().major()` as it still needs to use Java 8. This stuff is target for the RHSSO 7.6.X and hence Wildfly and Java8 support still needs to be kept. But since this is in the Wildfly module, it will be auto-removed once we are able to remove Wildfly distribution entirely.

pedroigor commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusPlatform.java:
```diff
@@ -159,4 +166,48 @@ public File getTmpDirectory() {
     }
+
+    @Override
+    public ClassLoader getScriptEngineClassLoader(Config.Scope scriptProviderConfig) {
```
In Quarkus we have a flat classpath so I'm unsure if we need to create a classloader?

mposolda replied:
@pedroigor You're right, Thanks for this point! After some digging, I figured that it is completely sufficient to just include the nashorn jar and it's dependencies directly inside "providers" directory. They are simply picked from there by the default flat classloader. So I changed the Quarkus part to just return null, which is simply using the same default (flat) classloader.

pedroigor reviewed (dismissed):
@mposolda Looking forward to have these changes in :)

mposolda commented:
@pedroigor @DGuhr Thanks guys! Will move forward with the testsuite update for this PR and then will "undraft" it and ask you for re-review. So there will be some additional changes, but possibly just in the testsuite part.

mposolda committed:
Script tests working with Java 17 on Wildfly and Quarkus

mposolda commented:
@pedroigor @DGuhr @kami619 I've removed the "draft" flag from this PR and I've updated testsuite part. Now the script tests should be working on JDK 17 on both Wildfly and Quarkus. Do you please have a chance to re-review this PR?

I've added testsuite changes in the latest commit for the easier re-review, but I am planning to squash the commits before merge. I am also planning to follow with the PRs for the quickstarts and documentation.

@kami619 In order to have tests passing on JDK 17, there is also need to prepare the servers with JDK 17 to have the necessary nashorn bits available on the servers. I've used this for my testing on local machine with Java 17:
```
cd testsuite/integration-arquillian/servers
mvn clean install -Pauth-server-wildfly
cd ../tests/base
mvn clean install -Pauth-server-wildfly -Dtest=ScriptAuthenticatorTest
```
Same should work also for `auth-server-quarkus` . And also it should hopefully work for all the other tests, which use javascript engine (protocol mappers, authorization), however I did not test other besides ScriptAuthenticatorTest.

kami619 commented:
@mposolda thank you for the detailed instructions, I shall pick this up and test it on my local.

mposolda commented:
@pedroigor Thanks Pedro!

@kami619 Ok, Thanks. I will wait for you to confirm if your tests are ok before merging this.

mabartos commented:
@mposolda After some discussion with @kami619, I was trying to execute the `ScriptAuthenticatorTest` test with JDK11 and JDK17. For JDK11 everything works as expected, but for JDK17 there are some issues. It seems the Script authenticator doesn't work for me. I'll try to investigate it a little bit.

mposolda commented:
@mabartos Hmm.... This is strange.

I've just checked again in this branch. I've rebuild it with Java 11 (from the top codebase directory):
```
mvn clean install -DskipTests=true -Pdistribution
```

And then executed this with Java 17:
```
cd testsuite/integration-arquillian/servers
mvn clean install -Pauth-server-wildfly
cd ../tests/base
mvn clean install -Pauth-server-wildfly -Dtest=ScriptAuthenticatorTest
```

Last command works fine also with `-Dtest=org.keycloak.testsuite.authz.AuthorizationTest` (The test which @kami619 pointed to me this morning).

Do you see different results with the command sequence like this? If not, can you check if in the ZIP file `testsuite/integration-arquillian/servers/auth-server/jboss/wildfly/integration-arquillian-servers-auth-server-wildfly-999-SNAPSHOT.zip` you see the nashorn module inside `auth-server-wildfly/modules/system/layers/keycloak/org/openjdk/nashorn/nashorn-core/main` ?

mabartos reviewed (approved):
@mposolda Oh, sorry, I tried to repeat those steps again and it seems everything's fine even for the WF. Sorry for the confusion. The module is present there.

Manually tested with Open JDK11 and Open JDK17 and everything works as expected.

kami619 commented:
@mabartos thanks for the confirmation. @mposolda lets merge it if its all good from your end as martin was able to verify it for us

mposolda commented:
@mabartos @kami619 Thanks guys!

BenjaminVettori commented:
I am uncertain if all dependencies of `nashorn-core.jar` are necessary, but when i looked them up at maven repository, i found those:

```
<dependencies>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm</artifactId>
      <version>7.3.1</version>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-commons</artifactId>
      <version>7.3.1</version>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-tree</artifactId>
      <version>7.3.1</version>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-util</artifactId>
      <version>7.3.1</version>
    </dependency>
  </dependencies>
```

seems like `asm-tree` was not mentioned before.
I added nashorn-core.jar (15.4) and the four dependency jar files (7.3.1) and then the default javascript policy started to work.

Just a small side-note: A Default (javascript) policy is generated, when you enable fine grained permissions for a client, even if the javascript dependency is missing. If you try to evaluate permissions with the default policy, there happens just nothing in the gui (no error, just a 500 response from the server in the developer tools). You have to look up server logs to find the root cause.

However, once you add the missing jar files to the provider folder and run `kc.sh build`, everything seems to work.
